lib/prompb/prompb.go

```diff
@@ -102,57 +227,93 @@ func (wr *WriteRequest) UnmarshalProtobuf(src []byte) (err error) {
 	wr.Timeseries = tss
 	wr.labelsPool = labelsPool
 	wr.samplesPool = samplesPool
+	wr.exemplarsPool = exemplarsPool
+	wr.histogramsPool = histogramsPool
 	return nil
 }
 
-func (ts *TimeSeries) unmarshalProtobuf(src []byte, labelsPool []Label, samplesPool []Sample) ([]Label, []Sample, error) {
+func (ts *TimeSeries) unmarshalProtobuf(src []byte, labelsPool *[]Label, samplesPool *[]Sample, exemplarsPool *[]Exemplar, histogramsPool *[]Histogram) error {
```

Comment:
Lets  group pool object into one struct `sharedPool` and use it instead of explicit reference/deference of pointers. From reader point of view its easier to understand the code.
```go
type sharedPool struct {
 labelsPool  []Label
 samplesPool []Sample
 exemplarsPool []Exemplar
  histogramsPool []Histogram
}

func (ts *TimeSeries) unmarshalProtobuf(src []byte, sp *sharedPool) error {
        labelsPoolLen := len(sp.labelsPool)
	samplesPoolLen := len(sp.samplesPool)
	exemplarsPoolLen := len(sp.exemplarsPool)
	histogramsPoolLen := len(sp.histogramsPool)
       ....
 	ts.Labels = sp.labelsPool[labelsPoolLen:]
	ts.Samples = sp.samplesPool[samplesPoolLen:]
	ts.Exemplars = sp.exemplarsPool[exemplarsPoolLen:]
	ts.Histograms = sp.histogramsPool[histogramsPoolLen:]
}
```

Reply:
updated as well